In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

In [2]:
# Load the CSV file
file_path = 'resources/board_games.csv'  
df = pd.read_csv(file_path)

In [3]:
# Display the first few rows and the summary of the DataFrame
print(df.columns)
df.head()

Index(['game_id', 'description', 'image', 'max_players', 'max_playtime',
       'min_age', 'min_players', 'min_playtime', 'name', 'playing_time',
       'thumbnail', 'year_published', 'artist', 'category', 'compilation',
       'designer', 'expansion', 'family', 'mechanic', 'publisher',
       'average_rating', 'users_rated'],
      dtype='object')


game_id                                        description  \
0        1  Die Macher is a game about seven sequential po...   
1        2  Dragonmaster is a trick-taking card game based...   
2        3  Part of the Knizia tile-laying trilogy, Samura...   
3        4  When you see the triangular box and the luxuri...   
4        5  In Acquire, each player strategically invests ...   

                                          image  max_players  max_playtime  \
0   //cf.geekdo-images.com/images/pic159509.jpg            5           240   
1   //cf.geekdo-images.com/images/pic184174.jpg            4            30   
2  //cf.geekdo-images.com/images/pic3211873.jpg            4            60   
3   //cf.geekdo-images.com/images/pic285299.jpg            4            60   
4   //cf.geekdo-images.com/images/pic342163.jpg            6            90   

   min_age  min_players  min_playtime            name  playing_time  ...  \
0       14            3           240      Die Macher           240  ...   
1       12            3            30    Dragonmaster            30  ...   
2       10            2            30         Samurai            60  ...   
3       12            2            60  Tal der Könige            60  ...   
4       12            3            90         Acquire            90  ...   

                        artist                        category compilation  \
0          Marcus Gschwendtner  Economic,Negotiation,Political         NaN   
1                   Bob Pepper               Card Game,Fantasy         NaN   
2              Franz Vohwinkel      Abstract Strategy,Medieval         NaN   
3                          NaN                         Ancient         NaN   
4  Scott Okumura,Peter Whitley                        Economic         NaN   

                designer expansion  \
0     Karl-Heinz Schmiel       NaN   
1  G. W. "Jerry" D'Arcey       NaN   
2          Reiner Knizia       NaN   
3      Christian Beierer       NaN   
4            Sid Sackson       NaN   

                                              family  \
0         Country: Germany,Valley Games Classic Line   
1                                   Animals: Dragons   
2  Asian Theme,Country: Japan,Knizia tile-laying ...   
3             Country: Egypt,Promotional Board Games   
4                                3M Bookshelf Series   

                                            mechanic  \
0  Area Control / Area Influence,Auction/Bidding,...   
1                                       Trick-taking   
2  Area Control / Area Influence,Hand Management,...   
3  Action Point Allowance System,Area Control / A...   
4       Hand Management,Stock Holding,Tile Placement   

                                           publisher average_rating  \
0  Hans im Glück Verlags-GmbH,Moskito Spiele,Vall...        7.66508   
1                           E.S. Lowe,Milton Bradley        6.60815   
2  999 Games,ABACUSSPIELE,Astrel Games,Ceilikan J...        7.44119   
3                                             KOSMOS        6.60675   
4  3M,Avalon Hill,Avalon Hill (Hasbro),Dujardin,G...        7.35830   

  users_rated  
0        4498  
1         478  
2       12019  
3         314  
4       15195  

[5 rows x 22 columns]

In [4]:
# Check for missing values
missing_values = df.isnull().sum()
print(missing_values[missing_values > 0])


image              1
thumbnail          1
artist          2773
category          94
compilation    10122
designer         126
expansion       7780
family          2808
mechanic         950
publisher          3
dtype: int64


In [5]:
# Remove duplicates
df = df.drop_duplicates()

# Check the number of rows before and after
print("Number of rows after removing duplicates:", len(df))


Number of rows after removing duplicates: 10532


In [6]:
# Example: Standardize categorical variables (e.g., lowercase the 'category' column)
df['category'] = df['category'].str.lower()

# Display unique categories to confirm changes
print(df['category'].unique())
df['category'].iloc[0]

['economic,negotiation,political' 'card game,fantasy'
 'abstract strategy,medieval' ... 'bluffing,deduction,dice'
 'fantasy,fighting,miniatures,sports'
 'bluffing,horror,maze,movies / tv / radio theme,science fiction']


'economic,negotiation,political'

In [7]:
# Example: Standardize categorical variables (e.g., lowercase the 'category' column)
df['mechanic'] = df['mechanic'].str.lower()

# Display unique categories to confirm changes
print(df['mechanic'].unique())

['area control / area influence,auction/bidding,dice rolling,hand management,simultaneous action selection'
 'trick-taking'
 'area control / area influence,hand management,set collection,tile placement'
 ...
 'action / movement programming,card drafting,co-operative play,dice rolling,modular board,variable player powers'
 'area control / area influence,auction/bidding,stock holding,worker placement'
 'action / movement programming,co-operative play,modular board,simultaneous action selection']


In [8]:
# Check data types
print(df.dtypes)

game_id             int64
description        object
image              object
max_players         int64
max_playtime        int64
min_age             int64
min_players         int64
min_playtime        int64
name               object
playing_time        int64
thumbnail          object
year_published      int64
artist             object
category           object
compilation        object
designer           object
expansion          object
family             object
mechanic           object
publisher          object
average_rating    float64
users_rated         int64
dtype: object


In [9]:
# Drop columns that are not needed or have excessive missing values
columns_to_drop = ['image', 'thumbnail', 'compilation', 'year_published', 'artist', 'designer', 'publisher', 'family']
df.drop(columns=columns_to_drop, inplace=True)
# df=df.dropna().reset_index(drop=True)
# Check the DataFrame to confirm the columns have been dropped
print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10532 entries, 0 to 10531
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   game_id         10532 non-null  int64  
 1   description     10532 non-null  object 
 2   max_players     10532 non-null  int64  
 3   max_playtime    10532 non-null  int64  
 4   min_age         10532 non-null  int64  
 5   min_players     10532 non-null  int64  
 6   min_playtime    10532 non-null  int64  
 7   name            10532 non-null  object 
 8   playing_time    10532 non-null  int64  
 9   category        10438 non-null  object 
 10  expansion       2752 non-null   object 
 11  mechanic        9582 non-null   object 
 12  average_rating  10532 non-null  float64
 13  users_rated     10532 non-null  int64  
dtypes: float64(1), int64(8), object(5)
memory usage: 1.1+ MB
None


In [10]:
# Check for missing values before filling
remaining_missing_values = df.isnull().sum()
print(remaining_missing_values[remaining_missing_values > 0])


category       94
expansion    7780
mechanic      950
dtype: int64


In [11]:
# Category counts
df['category_count'] = df['category'].apply(lambda x: len(str(x).split(',')))

In [12]:
# Mechanic counts
df['mechanic_count'] = df['mechanic'].apply(lambda x: len(str(x).split(',')))

In [13]:
# Replace NaN with None in the 'expansion' column
df['expansion'] = df['expansion'].replace({np.nan: None})

# Create a new column 'has_expansion' that says 'Yes' if 'expansion' has a value, otherwise 'No'
df['has_expansion'] = df['expansion'].apply(lambda x: 1 if x is not None else 0)

# Display the dataframe to check the changes
print(df[['expansion', 'has_expansion']].head())

  expansion  has_expansion
0      None              0
1      None              0
2      None              0
3      None              0
4      None              0


In [14]:
df.has_expansion.value_counts()

has_expansion
0    7780
1    2752
Name: count, dtype: int64

In [15]:
df=df.drop(['expansion'], axis=1)
df=df.dropna().reset_index(drop=True)

In [16]:
# Check the unique values in the 'category' column
unique_categories = df['category'].unique()

# To print each unique category
for category in unique_categories:
    print(category)

economic,negotiation,political
card game,fantasy
abstract strategy,medieval
ancient
economic
civilization,nautical
abstract strategy
civilization,fantasy
exploration
fantasy,travel
card game,farming,negotiation
ancient,mythology
negotiation
bluffing,negotiation,science fiction
collectible components,dice,fighting,print & play
miniatures,racing,science fiction
american west,bluffing,city building
science fiction
adventure,exploration,fantasy
fantasy,political,wargame
civilization,negotiation,political,science fiction,space exploration,wargame
civilization,economic,medieval,renaissance
economic,political,wargame
card game,humor,negotiation,political
fighting,humor
adventure,electronic,exploration,fantasy,fighting
adventure,exploration,fantasy,fighting,miniatures
adventure,horror,novel-based
science fiction,wargame
bluffing,civilization,fantasy,negotiation,political
dice
abstract strategy,ancient,civilization,territory building
aviation / flight,economic
card game
renaissance
city buildin

In [17]:
df.category.value_counts()

category
wargame,world war ii                                                                            446
card game                                                                                       381
abstract strategy                                                                               239
napoleonic,wargame                                                                              121
economic                                                                                        112
                                                                                               ... 
arabian,medieval,political,wargame                                                                1
card game,collectible components,fantasy,horror,novel-based,science fiction,video game theme      1
action / dexterity,humor,party game,trivia                                                        1
abstract strategy,animals,memory                                                           

In [18]:
df.head()

game_id                                        description  max_players  \
0        1  Die Macher is a game about seven sequential po...            5   
1        2  Dragonmaster is a trick-taking card game based...            4   
2        3  Part of the Knizia tile-laying trilogy, Samura...            4   
3        4  When you see the triangular box and the luxuri...            4   
4        5  In Acquire, each player strategically invests ...            6   

   max_playtime  min_age  min_players  min_playtime            name  \
0           240       14            3           240      Die Macher   
1            30       12            3            30    Dragonmaster   
2            60       10            2            30         Samurai   
3            60       12            2            60  Tal der Könige   
4            90       12            3            90         Acquire   

   playing_time                        category  \
0           240  economic,negotiation,political   
1            30               card game,fantasy   
2            60      abstract strategy,medieval   
3            60                         ancient   
4            90                        economic   

                                            mechanic  average_rating  \
0  area control / area influence,auction/bidding,...         7.66508   
1                                       trick-taking         6.60815   
2  area control / area influence,hand management,...         7.44119   
3  action point allowance system,area control / a...         6.60675   
4       hand management,stock holding,tile placement         7.35830   

   users_rated  category_count  mechanic_count  has_expansion  
0         4498               3               5              0  
1          478               2               1              0  
2        12019               2               4              0  
3          314               1               4              0  
4        15195               1               3              0

In [19]:
# Drop missing values
df = df.dropna().reset_index(drop=True)  # Drop rows with any NaN values
df = df.dropna(axis=1).reset_index(drop=True)  # Drop columns with any NaN values
# Check for NaN values in each column
print(df.isna().sum())

game_id           0
description       0
max_players       0
max_playtime      0
min_age           0
min_players       0
min_playtime      0
name              0
playing_time      0
category          0
mechanic          0
average_rating    0
users_rated       0
category_count    0
mechanic_count    0
has_expansion     0
dtype: int64


In [20]:
# Assuming you have a DataFrame named 'df' and 'playing_time' is one of the columns
# Let's get the unique values of the 'playing_time' column to understand its distribution.

playing_time_unique = df['playing_time'].unique()
playing_time_unique


array([  240,    30,    60,    90,    20,   120,    45,    25,     5,
         360,   340,   180,   300,    40,   150,   480,    15,    75,
         200, 12000,     0,  2480,    70,    10,  1200,   100,   210,
          80,  6000,  1440,   270,    35,   420,  4320,   960,  3600,
       60000,   720,     6,  5400,  2880,  1500,   700,  3000,   600,
           1,    12,  4500,  1800,    23,    50,  2400,   105,   160,
         900,     2,     3,    55,   222,   135, 17280,   540,   500,
          13,   400,   140,  1000,   108,    39,   750,    95,    68,
          32,   125,   165,   290,     4,   115,   130,    42,    11,
          19], dtype=int64)

In [21]:
print(df.columns)


Index(['game_id', 'description', 'max_players', 'max_playtime', 'min_age',
       'min_players', 'min_playtime', 'name', 'playing_time', 'category',
       'mechanic', 'average_rating', 'users_rated', 'category_count',
       'mechanic_count', 'has_expansion'],
      dtype='object')


In [22]:
# Define the function to bin playtime
def bin_playtime(time):
    if time <= 30:
        return 'Short (0-30 minutes)'
    elif 31 <= time <= 60:
        return 'Medium (31-60 minutes)'
    elif 61 <= time <= 120:
        return 'Long (61-120 minutes)'
    elif 121 <= time <= 240:
        return 'Very Long (121-240 minutes)'
    else:
        return 'Extended (241+ minutes)'

# Apply the function to the 'playing_time' column
df['binned_playtime'] = df['playing_time'].apply(bin_playtime)

# Display the binned playtime counts
binned_playtime_counts = df['binned_playtime'].value_counts()
print(binned_playtime_counts)

binned_playtime
Short (0-30 minutes)           3669
Medium (31-60 minutes)         2768
Long (61-120 minutes)          1807
Very Long (121-240 minutes)     856
Extended (241+ minutes)         407
Name: count, dtype: int64


In [23]:
# Define the mechanics bins
mechanics_bin = {
    'Action & Movement Mechanics': [
        'action / movement programming',
        'area movement',
        'grid movement',
        'point to point movement',
        'roll / spin and move',
        'route/network building'
    ],
    'Strategic Mechanics': [
        'area control / area influence',
        'auction/bidding',
        'betting/wagering',
        'commodity speculation',
        'partnerships',
        'player elimination',
        'press your luck',
        'take that',
        'trading',
        'voting'
    ],
    'Cooperative & Social Mechanics': [
        'co-operative play',
        'campaign / battle card driven',
        'storytelling',
        'singing'
    ],
    'Game Structure Mechanics': [
        'area enclosure',
        'modular board',
        'time track',
        'variable phase order',
        'variable player powers'
    ],
    'Card & Pool Mechanics': [
        'card drafting',
        'deck / pool building',
        'set collection',
        'trick-taking'
    ],
    'Randomness & Chance Mechanics': [
        'dice rolling',
        'chit-pull system',
        'memory',
        'secret unit deployment'
    ],
    'Pattern & Design Mechanics': [
        'crayon rail system',
        'line drawing',
        'pattern building',
        'pattern recognition',
        'tile placement',
        'simulation'
    ],
    'Unique Mechanics': [
        'action point allowance system',
        'hex-and-counter',
        'paper-and-pencil'
    ],
    'Others': [
        'simultaneous action selection',
        'stock holding'
    ]
}
# Function to bin mechanics
def bin_mechanics(mechanics):
    # Split mechanics into a list if they're comma-separated
    mechanics_list = [m.strip() for m in mechanics.split(',')]
    for category, mechanics_group in mechanics_bin.items():
        if any(mech in mechanics_list for mech in mechanics_group):
            return category
    return 'Other'  # Default category if no match is found

# Apply the function to create a new column 'binned_mechanics'
df['binned_mechanics'] = df['mechanic'].apply(bin_mechanics)


In [24]:
# Define bins and labels for kids' games
bins = [-1, 3, 6, 9, 12, 18]  # Age ranges
labels = ['Toddler (0-3)', 'Early Childhood (4-6)', 'Middle Childhood (7-9)', 
          'Pre-Teen (10-12)', 'Teen (13+)']

# Create a new column for binned ages
df['binned_min_age'] = pd.cut(df['min_age'], bins=bins, labels=labels, right=True)

# Display the DataFrame with the new binned column
print(df[['min_age', 'binned_min_age']].head())

   min_age    binned_min_age
0       14        Teen (13+)
1       12  Pre-Teen (10-12)
2       10  Pre-Teen (10-12)
3       12  Pre-Teen (10-12)
4       12  Pre-Teen (10-12)


In [25]:
# Your existing categories_bin dictionary
categories_bin = {
    "History/War": ["Ancient", "World War I", "World War II", "American Revolutionary War", "Vietnam War", 
                    "American Civil War", "American Indian Wars", "Napoleonic", "Post-Napoleonic", "Korean War", 
                    "American West", "Age of Reason", "Pike and Shot", "Prehistoric"],
    
    "Fantasy/Adventure": ["Fantasy", "Mythology", "Adventure", "Exploration", "Pirates", "Medieval"],
    
    "Science Fiction/Futuristic": ["Science Fiction", "Space Exploration", "Zombies", "Video Game Theme"],
    
    "Strategy": ["Wargame", "Abstract Strategy", "Political", "Civilization", "Territory Building", 
                 "Transportation", "Environmental", "Industry / Manufacturing", "Economic", "Trains"],
    
    "Educational/Intellectual": ["Math", "Educational", "Number", "Puzzle", "Trivia"],
    
    "Social/Party": ["Party Game", "Bluffing", "Negotiation", "Murder/Mystery", "Mafia", "Word Game"],
    
    "Entertainment/Pop Culture": ["Movies / TV / Radio theme", "Comic Book / Strip", "Novel-based", "Book", 
                                  "Humor", "Music"],
    
    "Children's/Family": ["Children's Game", "Memory", "Dice", "Expansion for Base-game", "Collectible Components"],
    
    "Real-life/Simulation": ["Farming", "Racing", "City Building", "Modern Warfare", "Aviation / Flight", "Medical", 
                             "Transportation", "Nautical", "Travel"],
    
    "Mystery/Crime": ["Spies/Secret Agents", "Murder/Mystery", "Deduction"],
    
    "Horror": ["Horror", "Zombies", "Murder/Mystery"],
    
    "Religion/Mythology": ["Religious", "Mythology"],
    
    "Unknown/Miscellaneous": ["Unknown", "Action / Dexterity", "Maze", "Electronic", "Game System", 
                              "Real-time", "Print & Play", "Mature / Adult", "Card Game", "Arabian"]
}

def get_categories(row):
    # Normalize and split the category string
    categories = [cat.strip().lower() for cat in row['category'].split(',')]
    
    # Check each genre against the categories in categories_bin
    for genre in categories:
        for category, keywords in categories_bin.items():
            if genre in [k.lower() for k in keywords]:  # Check for match, case insensitive
                return category  # Return the first matched category
    
    return 'Other'  # Return 'Other' if no matches are found

# Apply the updated function
df['binned_category'] = df.apply(get_categories, axis=1)


In [26]:
# Check the distribution of binned categories
print(df['binned_category'].value_counts())

binned_category
Unknown/Miscellaneous         2508
Strategy                      1976
Fantasy/Adventure             1047
History/War                    911
Children's/Family              759
Social/Party                   725
Real-life/Simulation           692
Entertainment/Pop Culture      258
Science Fiction/Futuristic     229
Mystery/Crime                  154
Other                           85
Educational/Intellectual        83
Horror                          76
Religion/Mythology               4
Name: count, dtype: int64


In [27]:
# Save the cleaned DataFrame to a CSV file in the resources folder
df.to_csv('resources/clean.csv', index=False)


# Summary Statistics
Numerical Columns:
Count: All columns have 10,532 entries.
Mean Values:
max_players: 5.66
max_playtime: 91.34 minutes
min_age: 9.71 years
min_players: 2.07
min_playtime: 80.88 minutes
playing_time: 91.34 minutes
average_rating: 6.37
users_rated: 870.08
Standard Deviations: Varying values across columns, indicating diverse data distribution.
Min/Max Values:
Players: 0 to 999
Playtime: 0 to 60,000 minutes
Rating: 1.38 to 9.00
Users Rated: 50 to 67,655
Unique Values
Description: Various detailed game descriptions, with some highlighting gameplay mechanics, history, and components.
Name: A diverse list of game titles including "Die Macher", "Dragonmaster", and "Robo Rally (2016)".
Category: Unique categories like "economic, negotiation, political" and "card game, fantasy".
Expansion: Listings of expansions with many games having multiple expansions or none.